### PR TITLE
introduce entity that will keep info about services current state and integrate it into DASer and Syncer

### LINKED ISSUE #724: service: Add flag for stopped services
We should add a flag that will prevent the service from being stopped twice.

#Why
Currently, all our Stop calls depends on fx.Lifecycle which guarantees it will be called once. Receiving of BEFP also cause service stopping. The second Stop call can either crash or freeze the node.

### PR DESCRIPTION
**Pull request by vgonkivs**
Resolves #724

### CODE REVIEW

File: libs/utils/state_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package utils
+
+import "sync/atomic"
+
+type state uint32
+
+const (
+	Stopped state = iota
+	Runnnig
```
Comment (renaynay) [inline libs/utils/state_wrapper.go:9]:
```suggestion
	Running
```

File: libs/utils/state_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package utils
```
Comment (Wondertan) [inline libs/utils/state_wrapper.go:1]: * `libs/utils` does not make sense(even though I made it, but I don't like this) 
* If this is needed for service management, we should make a `service` pkg.

File: libs/utils/state_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+type state uint32
+
+const (
+	Stopped state = iota
+	Runnnig
+)
+
+// StateWrapper provides an interface to change an check service state
+type StateWrapper interface {
```
Comment (Wondertan) [inline libs/utils/state_wrapper.go:13]: StateWrapper is a wrong name for the interface, IMO. Also, do we really need this interface? Do you expect multiple implementations of it?  What about a simple Service struct that all our services embed through composition?

File: das/daser.go [hunk 4]
```diff
@@ -82,6 +87,11 @@ func (d *DASer) Start(context.Context) error {
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
+	if d.state.GetState() == utils.Stopped {
+		log.Debug("Daser is stopped")
+		return nil
+	}
+	d.state.SetState(utils.Stopped)
```
Comment (Wondertan) [inline das/daser.go:94]: Should be in the end of the method

File: libs/utils/state_wrapper.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	Stopped state = iota
+	Running
+)
+
+// StateWrapper provides an interface to change an check service state
+type StateWrapper interface {
+	// SetState allows to change service state
+	SetState(state)
+	// GetState returns service current state
+	GetState() state
```
Comment (Wondertan) [inline libs/utils/state_wrapper.go:17]: You can omit Get and make it just State

**Review by Wondertan — COMMENTED** [review]
Making Service like this `libs/service.Service` and usage like :

```go
type Syncer struct{
	service.Service
	...
}

func (s *Syncer) Stop(context.Context) error {
	if s.state.State() == service.Stopped {
		return nil
	}
}
```
WDYT?

**Comment by Wondertan** [comment]
Other thoughts:
* As we discussed, this should be part of FX, and FX can track on its own the state of each Service that was started or not. No matter how we try to solve this problem, it will be better solved by the uber-go/fx and their team. We need deeper integration with the dependency graph to not just stop a Service but also stop everything that depends on the stopped service. We should make a proposal as I did in https://github.com/uber-go/fx/issues/825 for this.
* @vgonkivs, can you tell me if you have other plans for this structure? It feels like it is a bit of overengineering for a simple atomic switch. If there are other long term plans/ideas on how we can use this, then I am fine to keep it, but if not, I see no reason to keep it, considering the proper solution should be done on the FX side.

**Comment by vgonkivs** [comment]
@Wondertan, no other plans except provided. I do not think that it's overengineering. It allows to re-use this structure in different services and provides simple aliases for On/Off state

File: libs/utils/state_wrapper.go (hunk 3 above)
Comment (vgonkivs) [reply libs/utils/state_wrapper.go]: removed

File: libs/utils/state_wrapper.go (hunk 2 above)
Comment (vgonkivs) [reply libs/utils/state_wrapper.go]: done

**Comment by Wondertan** [comment]
Can you also rename the PR to something more appropriate? We already have a state service that accesses the State of the blockchain

File: das/daser.go [hunk 6]
```diff
@@ -82,6 +86,11 @@ func (d *DASer) Start(context.Context) error {
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
+	if d.State() == service.Stopped {
+		log.Debug("Daser is stopped")
```
Comment (adlerjohn) [inline das/daser.go:90]:
```suggestion
		log.Debug("DASer is stopped")
```

File: libs/service/service.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+type State uint32
+
+const (
+	Stopped State = iota
+	Running
+)
+
+// Service provides an interface to change an check service state
+type Service struct {
+	state uint32
```
Comment (adlerjohn) [inline libs/service/service.go:14]: ?
```suggestion
	state State
```

Comment (vgonkivs) [reply libs/service/service.go:14]: In this case storing and loading state will be more complex:
```
* atomic.StoreUint32((*uint32)(&s.state), uint32(state))
* return State(atomic.LoadUint32((*uint32)(&s.state)))
```

IMO, better to leave `uint32`

Comment (adlerjohn) [reply libs/service/service.go:14]: Then maybe the naming of `state` here is inconsistent? `s.state` seems to be a key for the state, not a state? If so, it it's 1) not clearly named or documented, and 2) should be `uint64`, unless there's a very specific reason it must be a `uint32`.

**Comment by vgonkivs** [comment]
Closing this PR in favour of more priority tasks.

**Pull request closed without merging**